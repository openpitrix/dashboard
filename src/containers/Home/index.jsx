import React, { Component } from 'react';
import { observer, inject } from 'mobx-react';

import Nav from 'components/Nav';
import AppList from 'components/AppList';
import styles from './index.scss';

@inject('rootStore')
@observer
export default class Home extends Component {
  static async onEnter({ appStore }) {
    await appStore.fetchApps();
  }

  render() {
    const { config, appStore } = this.props.rootStore;

    return (
      <div className={styles.home}>
        <Nav className={styles.nav} navs={config.navs}/>
        <AppList className={styles.apps} apps={appStore.apps}/>
      </div>
    );
  }
}
