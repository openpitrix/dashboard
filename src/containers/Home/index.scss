@import "~scss/variables";

.home{
  display: table;
  width: 100%;
  height: calc(100vh - #{$footer-height});
  .nav {
    display: table-cell;
    width: 210px;
    padding-left: calc(50vw - #{$content-width}/2);
    border-right: 1px solid $border;
    background: #ffffff;
    box-sizing: content-box;
    vertical-align: top;
  }
  .apps {
    display: table-cell;
    vertical-align: top;
    padding-right: calc(50vw - #{$content-width}/2);
  }
}