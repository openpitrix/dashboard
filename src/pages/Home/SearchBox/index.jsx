import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withRouter } from 'react-router-dom';
import { translate } from 'react-i18next';
import { observer, inject } from 'mobx-react';

import { Input } from 'components/Base';
import { getUrlParam } from 'utils/url';

import styles from './index.scss';

@translate()
@inject(({ rootStore }) => ({
  rootStore,
  appStore: rootStore.appStore
}))
@observer
export class SearchBox extends React.Component {
  static propTypes = {
    className: PropTypes.string
  };

  onSearch = value => {
    this.props.rootStore.setSearchWord(value);
    this.props.history.push(`/?q=${value}`);
  };

  onClearSearch = () => {
    const { rootStore, appStore, history } = this.props;
    appStore.currentPage = 1;
    rootStore.setSearchWord('');
    history.push('/');
  };

  render() {
    const { rootStore, t, className } = this.props;
    const { searchWord } = rootStore;

    return (
      <Input.Search
        className={classnames(styles.search, className)}
        placeholder={t('search.placeholder')}
        value={searchWord || getUrlParam('q')}
        onSearch={this.onSearch}
        onClear={this.onClearSearch}
      />
    );
  }
}

export default withRouter(SearchBox);
