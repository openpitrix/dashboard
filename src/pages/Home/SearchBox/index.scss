@import '~scss/vars';

.search {
  width: 360px;
  margin-top: 20px;

  input {
    color: $N500;
    padding-left: 36px;
    border-radius: 16px;
    border: 1px solid transparent;
    background-color: $N10;
    transition: all .3s ease-in-out;

    &::placeholder {
      font-size: 14px;
      color: $N45;
      line-height: 1.0;
      letter-spacing: 0;
    }
    &:focus {
      border: solid 1px $P75 !important;
    }

    &:hover {
      border: solid 1px $N45;
    }
  }
  :global {
    .icon {
      color: $N65;
      opacity: 0.5;
      &:hover, &:focus {
        opacity: 1;
      }
    }
  }
}