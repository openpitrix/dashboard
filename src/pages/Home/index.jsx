import React, { Fragment, Component } from 'react';
import classnames from 'classnames';
import { observer, inject } from 'mobx-react';
import { get, find } from 'lodash';

import { Notification } from 'components/Base';
import Nav from 'components/Nav';
import Banner from 'components/Banner';
import AppList from 'components/AppList';
import Loading from 'components/Loading';
import { getScrollTop, getScrollBottom } from 'src/utils';

import styles from './index.scss';

@inject(({ rootStore }) => ({
  rootStore,
  categoryStore: rootStore.categoryStore,
  appStore: rootStore.appStore
}))
@observer
export default class Home extends Component {
  async componentWillMount() {
    const { rootStore, match } = this.props;
    const { category, search } = match.params;

    if (category || search) {
      rootStore.setNavFix(true);
    } else {
      rootStore.setNavFix(false);
    }
  }

  async componentDidMount() {
    const {
      rootStore, appStore, categoryStore, match
    } = this.props;
    const { category, search } = match.params;
    const filterParams = { status: 'active', noLimit: true };

    window.scroll({ top: 0 });
    await categoryStore.fetchAll();

    if (!(category || search)) {
      this.threshold = this.getThreshold();
      window.onscroll = this.handleScroll;
    }

    if (category) {
      filterParams.category_id = category;
    }
    if (search) {
      filterParams.search_word = search;
    }
    await appStore.fetchAll(filterParams);

    appStore.homeApps = appStore.apps.slice();
  }

  componentWillUnmount() {
    const { appStore } = this.props;

    window.onscroll = null;
    appStore.reset();
  }

  getThreshold() {
    const headerNode = document.querySelector('.header');
    const bannerNode = document.querySelector('.banner');
    if (headerNode && bannerNode) {
      return bannerNode.clientHeight - headerNode.clientHeight;
    }
    return 0;
  }

  handleScroll = async () => {
    const { rootStore } = this.props;
    if (this.threshold <= 0) {
      return;
    }

    // judge header fixed
    const fixNav = rootStore.fixNav;
    const scrollTop = getScrollTop();
    const needFixNav = scrollTop > this.threshold;
    if (needFixNav && !fixNav) {
      rootStore.setNavFix(true);
    } else if (!needFixNav && fixNav) {
      rootStore.setNavFix(false);
    }

    // load app data progressive by window scroll
    const { appStore, categoryStore } = this.props;
    const { categories } = categoryStore;
    const len = categories.length;
    if (len > 0 && categories[len - 1].appFlag) {
      return;
    }

    const initLoadNumber = parseInt((document.body.clientHeight - 720) / 250);
    if (len > 0 && !categories[0].appFlag && !appStore.isLoading) {
      await this.loadAppData(categories, initLoadNumber);
    }

    const scrollBottom = getScrollBottom();
    if (scrollBottom < 100 && !appStore.isProgressive) {
      await this.loadAppData(categories);
    }
  };

  loadAppData = async (categories, initLoadNumber) => {
    const { categoryStore, appStore } = this.props;

    for (let i = 0; i < categories.length; i++) {
      if (!categories[i].appFlag) {
        categoryStore.categories[i].appFlag = true;
        await appStore.fetchAll({
          limit: 6,
          status: 'active',
          category_id: categories[i].category_id,
          noLoading: true
        });
        const temp = categoryStore.categories[i];
        categoryStore.categories[i] = {
          apps: appStore.apps,
          ...temp
        };
        if (initLoadNumber) {
          if (appStore.apps.length === 0) {
            initLoadNumber++;
          } else if (initLoadNumber === i + 1) {
            break;
          }
        } else if (appStore.apps.length > 0) {
          break;
        }
      }
    }
  };

  render() {
    const {
      rootStore, appStore, categoryStore, match
    } = this.props;
    const { fixNav } = rootStore;
    const { homeApps, isLoading, isProgressive } = appStore;
    const categories = categoryStore.categories;

    const { category, search } = match.params;
    const showApps = category || search ? homeApps.slice() : homeApps.slice(0, 3);
    const isHomePage = match.path === '/';
    const categoryTitle = get(
      find(categories, { category_id: category }),
      'name',
      ''
    );

    return (
      <Fragment>
        {isHomePage && <Banner />}
        <Notification />
        <div
          className={classnames(styles.content, { [styles.fixNav]: fixNav })}
        >
          <Nav className={styles.nav} navs={categories.toJSON()} />
          <Loading isLoading={isLoading} className={styles.homeLoad}>
            <AppList
              className={styles.apps}
              apps={showApps}
              categoryApps={categories.toJSON()}
              categoryTitle={categoryTitle}
              appSearch={search}
              isLoading={isLoading}
            />
            {isProgressive && (
              <div className={styles.loading}>
                <div className={styles.loadOuter}>
                  <div className={styles.loader} />
                </div>
              </div>
            )}
          </Loading>
        </div>
      </Fragment>
    );
  }
}
