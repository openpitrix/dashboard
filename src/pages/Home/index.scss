@import '~scss/vars';

.content {
  position: relative;
  margin: 0 auto;
  min-height: $content-height;
  width: $content-width;

  .homeLoad {
    margin-left: $nav-width;
  }
  .nav {
    position: absolute;
    top: 0;
    left: 0;
    width: $nav-width;
    height: 100%;
  }
  .apps {
    margin-left: $nav-width;
    padding-left: 24px;
    padding-bottom: $page-bottom;
  }

  .progressive{
    position: static;
    height: 60px;
    .loadOuter{
      position: static;
      padding: 30px 0;
      .loader {
        margin: 0 auto;
      }
    }
  }
  .loading {
    margin-top: -20px;
    margin-left: $nav-width;
    text-align: center;
    .loadOuter{
      display: inline-block;
      .loader {
        border-radius:50%;
        width:40px;
        height:40px;
        &:after {
          border-radius:50%;
          width:40px;
          height:40px;
        }
        border-top: 2px solid rgba($purple-light,0.2);
        border-right: 2px solid rgba($purple-light,0.2);
        border-bottom: 2px solid rgba($purple-light,0.2);
        border-left: 2px solid rgba($purple-light,1);
        transform: translateZ(0);
        animation:rotate 1.1s infinite linear;
      }
    }
  }
}

.fixNav {
  div[class*='_nav'] {
    position: fixed;
    top: $header-height;
    left: auto;
    height: $main-height;
  }

  div[class*='appList'] {
    padding-top: $header-height;
  }

  div[class*='_cateTitle']{
    padding-top: 72px;
  }
}
