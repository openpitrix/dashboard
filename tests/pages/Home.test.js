import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { observable, useStrict } from 'mobx';
import { Provider } from 'mobx-react';

import Home from 'containers/Home';

describe('Home', () => {
  let rootStore;

  beforeEach(() => {
    // turn off strict mode when testing with mock store
    useStrict(false);

    rootStore = observable({
      config: {},
      appStore: {}
    });
  });

  afterEach(() => {
    rootStore = null;
  });

  it('basic render', () => {
    const wrapper = mount(
      <Provider rootStore={rootStore}>
        <Home />
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('call onEnter', () => {
    // todo
  });
});
