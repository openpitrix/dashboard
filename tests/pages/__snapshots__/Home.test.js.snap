// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home basic render 1`] = `
<Provider
  rootStore={
    Object {
      "appStore": Object {},
      "config": Object {},
    }
  }
>
  <inject-Home-with-rootStore>
    <Home
      rootStore={
        Object {
          "appStore": Object {},
          "config": Object {},
        }
      }
    >
      <Banner>
        <div
          className="banner banner"
        >
          <div
            className="wrapper"
          >
            <img
              alt=""
              className="banner-img-1"
              src="/assets/1-1.svg"
            />
            <img
              alt=""
              className="banner-img-2"
              src="/assets/1-2.svg"
            />
            <img
              alt=""
              className="banner-img-3"
              src="/assets/1-3.svg"
            />
            <div
              className="title"
            >
              Application Management Platform on Multi-Cloud Environment.
            </div>
            <Search
              className="search"
              onSearch={[Function]}
              placeholder="Search apps in Pitrix..."
            >
              <div
                className="inputGroup search"
              >
                <Icon
                  name="search"
                >
                  <i
                    className="icon icon-search"
                  />
                </Icon>
                <input
                  className="input"
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Search apps in Pitrix..."
                  type="text"
                  value=""
                />
              </div>
            </Search>
          </div>
        </div>
      </Banner>
      <div
        className="contentOuter"
      >
        <div
          className="content"
        >
          <Nav
            className="nav"
          >
            <div
              className="nav nav"
            />
          </Nav>
          <AppList
            className="apps"
            fold={false}
          >
            <div
              className="appList apps"
            >
              <CardTitle
                more={true}
                title="Catalog Title"
              >
                <div
                  className="title"
                >
                  Catalog Title
                   
                  <span
                    className="more show"
                  >
                    more...
                  </span>
                </div>
              </CardTitle>
            </div>
          </AppList>
        </div>
      </div>
    </Home>
  </inject-Home-with-rootStore>
</Provider>
`;
